tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc

```diff
@@ -1046,12 +1046,39 @@ std::vector<int> LayoutAgnosticOpTransposer::GetVariadic4DFaninPorts(
 Status DefaultLayoutAgnosticOpTransposer::TransposeNode(
     TransposeContext* context, utils::MutableNodeView* node) {
   DCHECK(IsDefaultLayoutAgnosticOp(*node->node()));
-  if (!ShouldProcess(*context, *node) || !IsFanoutPortRankN(*node, 0, 4) ||
-      !IsAfterDstToSrcTransform(*context, *node)) {
+  const auto* output_shape_attr = node->GetAttr(kAttrOutputShape);
+  const auto& shape = output_shape_attr->list().shape(0);
+  const int rank = shape.dim_size();
+  std::string src_format = context->src_format;
+  std::string dst_format = context->dst_format;
+  // Update the format from 4D to 5D layout if necessary.
+  bool allow_5d = rank == 5 && (src_format == "NHWC" || src_format == "NCHW") &&
+                  (dst_format == "NHWC" || dst_format == "NCHW");
+  if (allow_5d) {
+    std::string src_format_3d = src_format == "NHWC" ? "NDHWC" : "NCDHW";
+    std::string dst_format_3d = dst_format == "NHWC" ? "NDHWC" : "NCDHW";
+    context->AssignDeviceAndDataFormats(context->target_device, src_format_3d,
+                                        dst_format_3d);
+  }
+  if (!ShouldProcess(*context, *node) || (rank != 4 && rank != 5) ||
```

Comment:
Is the `(rank != 4 && rank != 5)` check necessary? I think `IsFanoutPortRankN` should already cover this case?

Reply:
I think the `IsFanoutPortRankN` checks if the port x outputs rank N tensor, while `(rank != 4 && rank != 5)` checks if N is 4 or 5. So, for example, if the tensor dim is 3, we want the `(rank != 4 && rank != 5)` to return true to avoid layout optimization.

Reply:
In that case, then `!IsFanoutPortRankN(*node, 0, rank)` is not necessary from the looks of it, and can be removed? And can we move the `(rank != 4 && rank != 5)` to the beginning of the function, so we early terminate the function and not upgrade layouts?

Reply:
Oh, I see your point. The `rank` is obtained from output shape. So, yes, the `!IsFanoutPortRankN(*node, 0, rank)` seems unnecessary.

Reply:
Can the `(rank != 4 && rank != 5)` check be moved to near the beginning of `TransposeNode` (like right after getting `rank` at L1051)? That way, we can early terminate without checking and upgrading data formats.

Reply:
I think for the rank!=4 or !=5, the data format won't be changed because it is guarded by allow_5d.

I can make the change if you want, but I think it is unnecessary and putting all early-exit checkings in the single branch would be more clear.

Reply:
The branch is moved to L1051. PTAL.